torch/distributed/tensor/experimental/_attention.py

```diff
@@ -1456,3 +1574,162 @@ def set_rotate_method(rotate_method: str) -> None:
             "Context Parallel does not support "
             f"using {rotate_method} for kv shards rotation"
         )
+
+
+class ContextParallelMode(TorchDispatchMode):
+    def __init__(
+        self,
+        device_mesh: DeviceMesh,
+        block_mask: Optional[BlockMask] = None,
+        sharder: Optional[FlexAttentionSharder] = None,
+    ):
+        super().__init__()
+        self._device_mesh = device_mesh
+
+        # if user has not specified the flex_attention components
+        if block_mask is None or sharder is None:
+            self._cp_block_mask = None
+            self._sharder = None
+            self._sharding_map = torch.empty(0)
+            return
+
+        block_mask_tuple = block_mask.as_tuple()
+        assert len(block_mask_tuple) == 13
+        Q_LEN: int = block_mask_tuple[0]
+        KV_LEN: int = block_mask_tuple[1]
+
+        mask_mod: _mask_mod_signature = block_mask_tuple[-1]
+        Q_BLOCK_SIZE: int = block_mask_tuple[-3]
+        KV_BLOCK_SIZE: int = block_mask_tuple[-2]
+
+        # TODO: support other KV block sizes
+        assert Q_BLOCK_SIZE == KV_BLOCK_SIZE
+
+        # resolve CP device mesh info
+        assert device_mesh.ndim == 1
+        cp_rank = device_mesh.get_local_rank()
+        cp_group_size = device_mesh.size()
+        device_type = device_mesh.device_type
+
+        # the sharding function is configurable
+        sharding_map = sharder.gen_block_sharding_map(cp_group_size, block_mask).to(
```

Comment:
`_rewrite_mask_mod_for_cp()`, `_create_block_mask_cached()`,  `_cp_block_mask`, and `_sharding_map` should be the methods and attributes of `FlashAttentionSharder`.  `ContextParallelMode` should only keep `_sharder`. This can generalize `ContextParallelMode` be used for SDPA as well.